packages/api/src/perma-cache/put.js

```diff
@@ -0,0 +1,180 @@
+/* eslint-env serviceworker, browser */
+/* global Response */
+
+import { InvalidUrlError, TimeoutError, HTTPError } from '../errors.js'
+import { JSONResponse } from '../utils/json-response.js'
+import { normalizeCid } from '../utils/cid.js'
+
+/**
+ * @typedef {import('../env').Env} Env
+ * @typedef {Object} IpfsUrlParts
+ * @property {string} cid
+ * @property {string} url
+ * @property {string} path
+ */
+
+/**
+ * Handle perma-cache put request
+ *
+ * @param {Request} request
+ * @param {Env} env
+ */
+export async function permaCachePut(request, env) {
+  const sourceUrl = getSourceUrl(request, env)
+  const normalizedUrl = getNormalizedUrl(sourceUrl, env)
+  const r2Key = normalizedUrl.toString()
+
+  // Validate if we already have it in R2
+  let r2Object
+  try {
+    r2Object = await env.SUPERHOT.head(normalizedUrl.toString())
+  } catch (err) {
+    // TODO: R2 currently throws error here and does not follow types to return undefined.
+    // They say it will change soon to return undefined...
+    console.log(err)
+  }
+
+  if (!r2Object) {
+    // Fetch Response from provided URL
+    const response = await getResponse(request, env, normalizedUrl)
+    if (!response.ok) {
+      throw new HTTPError(await response.text(), 400)
+    }
+
+    // TODO: Validate headers per https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control
+    // Store in R2 and add to Database if not existent
+    r2Object = await env.SUPERHOT.put(r2Key, response.body, {
+      httpMetadata: response.headers,
+    })
+  }
+
+  // Store key with user namespace to handle concurrency and keeping track of all
+  const insertedAt = new Date().toISOString()
+  const kvKey = `${request.auth.user.id}/${r2Key}/${insertedAt}`
```

Comment:
Regarding reasoning, considering we have as requirements:
- track billing of given content length per time frame
- allow deletes

Considering we need to have a way of tracking if a user deletes and adds again a given URL to perma cache. This would result in two entries: one with deleted ts set and other with no deleted ts. So, we could bill and keep track of what content was stored within each timeframe.

Initially, I was expecting to simply use user id as key. But that will create a mess with race conditions on concurrent put/delete. So, relying on unique keys with prefix by userId is a clean solution that allows us to:
- Get list by prefix
- Check if user already added such URL to permacache (just list with prefix `userId/r2Key` and see if there is any entry with deleted undefined or just empty list...)
- Keep full history of transactions